File: frontend/src/api/runtimeTerminal.ts
import { apiFetch, getStoredAccessToken, resolveApiWebSocketUrl } from './client';

export type TerminalSessionCreateResponse = {
  session_id: string;
  deployment_id: string;
  service_id: string;
  status: string;
  websocket_path: string;
  expires_at: string;
  max_duration_seconds: number;
  idle_timeout_seconds: number;
  runtime_provider: string;
  message?: string | null;
};

export function createTerminalSession(deploymentId: string, serviceId: string) {
  return apiFetch<TerminalSessionCreateResponse>(
    `/deployments/${deploymentId}/runtime/services/${serviceId}/terminal`,
    { method: 'POST' },
  );
}

export function closeTerminalSession(sessionId: string) {
  return apiFetch<{ session_id: string; status: string; close_reason?: string | null }>(
    `/terminal-sessions/${sessionId}`,
    { method: 'DELETE' },
  );
}

export function terminalWebSocketUrl(websocketPath: string): string {
  const path = websocketPath.startsWith('/') ? websocketPath : `/${websocketPath}`;
  const token = getStoredAccessToken();
  const q = token ? `?token=${encodeURIComponent(token)}` : '';
  return resolveApiWebSocketUrl(`${path}${q}`);
}

File: backend/app/schemas/runtime_terminal.py
"""Schemas for interactive runtime terminal sessions."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class TerminalSessionCreateResponse(BaseModel):
    session_id: UUID
    deployment_id: UUID
    service_id: UUID = Field(description="Persisted runtime service resource row id")
    status: str
    websocket_path: str
    expires_at: datetime
    max_duration_seconds: int
    idle_timeout_seconds: int
    runtime_provider: str
    message: str | None = None


class TerminalSessionCloseResponse(BaseModel):
    session_id: UUID
    status: str
    close_reason: str | None = None
